packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -393,20 +447,42 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
   ) {
     final int? maxLength = this.maxLength;
 
-    if (maxLength == null || maxLength == -1 || newValue.text.characters.length <= maxLength)
+    if (maxLength == null ||
+      maxLength == -1 ||
+      newValue.text.characters.length <= maxLength) {
       return newValue;
+    }
 
     assert(maxLength > 0);
 
-    // If already at the maximum and tried to enter even more, keep the old
-    // value.
-    if (oldValue.text.characters.length == maxLength && !oldValue.composing.isValid) {
-      return oldValue;
+    switch (maxLengthEnforcement) {
+      case MaxLengthEnforcement.none:
+        return newValue;
+      case MaxLengthEnforcement.enforced:
+        // If already at the maximum and tried to enter even more, keep the old
+        // value.
+        if (oldValue.text.characters.length == maxLength) {
+          return oldValue;
```

Comment:
Currently 12345

Reply:
The case is solved with:
* If previous value has selection, new value can comes in with truncation.
* Else keep old value.

Reply:
One more edge case I want to think about: What if the user has set newValue programmatically, so there was never any selection?

Sorry to nitpick these edge cases :). I think it's fine if we keep the same behavior as before, I just want to make sure we think about it in case it has changed.

Reply:
Make sense, but I don't think that we handled this case before this change. Let me revert some commits to see how do we handle this before.